### PR TITLE
Enable handling of arbitrary if statements within actions for BMv2 back end

### PR DESCRIPTION
**Pull request by jafingerhut**

### CODE REVIEW

**Comment by fruffy** [comment]
This also fixes known failures in P4Testgen for STF and PTF:

https://github.com/p4lang/p4c/blob/main/backends/p4tools/modules/testgen/targets/bmv2/test/BMV2PTFXfail.cmake#L147

https://github.com/p4lang/p4c/blob/main/backends/p4tools/modules/testgen/targets/bmv2/test/BMV2STFXfail.cmake#L147

**Comment by fruffy** [comment]
Would this fix https://github.com/p4lang/p4c/issues/644? One of the oldest open issues for the compiler.

**Comment by jafingerhut** [comment]
@fruffy Would you like any additional reviews on this, or is it OK to merge?

**Review by fruffy — APPROVED** [review]
Tagging @smolkaj and @jonathan-dilorenzo as stakeholders. From my side this seems good-to-go. 

@r12f this might be useful for your DASH modelling efforts?

File: backends/bmv2/common/action.h [hunk 1]
```diff
@@ -25,7 +25,12 @@ namespace P4::BMV2 {
 class ActionConverter : public Inspector {
     ConversionContext *ctxt;
 
-    void convertActionBody(const IR::Vector<IR::StatOrDecl> *body, Util::JsonArray *result);
+    void convertActionBodyTop(const IR::Vector<IR::StatOrDecl> *body, Util::JsonArray *result);
+    void convertActionBody(const IR::Vector<IR::StatOrDecl> *body, Util::JsonArray *result,
```
Comment (fruffy) [inline backends/bmv2/common/action.h:29]: Nit: I'd consider making this a struct that has all this information. Makes it easier to change and read.

**Comment by jafingerhut** [comment]
> Tagging @smolkaj and @jonathan-dilorenzo as stakeholders. From my side this seems good-to-go.

Jonathan and Steffen -- I believe these changes are completely backwards compatible, meaning that if a P4 program for BMv2 v1model compiles and runs correctly today, it should continue to do so with these changes.

Call p4c-bm2-ss without these changes "A", and with these changes "B".

There exist P4 programs with if statements in actions that give no compile-time error with A, but generate BMv2 JSON files with bugs such that when they process packets, they behave differently than the P4 program specifies.  These are fairly rare -- I believe you must have nested if statements within an action body to trigger these bugs in version A, but they do exist.  Version B should fix all such bugs.

There also exist P4 programs with if statements in actions that give compile-time errors with A, and generate no BMv2 JSON file.  Such programs should compile without error with version B, and produce BMv2 JSON files that behave as specified by the P4 program.

File: backends/bmv2/common/action.h (hunk 1 above)
Comment (jafingerhut) [reply backends/bmv2/common/action.h:61]: Good idea.  Done in commit 7.

**Comment by r12f** [comment]
> Tagging @smolkaj and @jonathan-dilorenzo as stakeholders. From my side this seems good-to-go. 
> 
> 
> 
> @r12f this might be useful for your DASH modelling efforts? 

Woot!! Absolutely! Thanks a lot Fabian and Andy! Finally can remove the tricky ternary operators now!

**Comment by jafingerhut** [comment]
The only failing test is for compiling Tofono back end, which appears to be related to issues between not being able to recognize string listerals as type cstring, or something similar to that, unrelated to these changes.

**Comment by smolkaj** [comment]
Awesome! I have definitely wanted this in the past. cc @kheradmandG for visibility.

**Pull request merged**